### ISSUE TITLE
sovits训练完成后在SoVITS_weights下没有文件，推理的时候无法选择新模型，gpt训练完就有文件

### ISSUE THREAD

**Issue by cjszhj** [RVC-Boss/GPT-SoVITS#1197]
我用的是windows 整合包，不应该存在配置问题。在微调这一步，SoVITS和gpt训练都能正常完成。
显卡：1060 5G，参数也没设置很高。
sovits训练输出：
J:\AI\GPT-SoVITS-beta0217fix3\GPT-SoVITS-beta0217fix3\runtime\lib\site-packages\torch\autograd\__init__.py:200: UserWarning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.  This is not an error, but may impair performance.
grad.sizes() = [1, 9, 96], strides() = [80160, 96, 1]
bucket_view.sizes() = [1, 9, 96], strides() = [864, 96, 1] (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\torch\csrc\distributed\c10d\reducer.cpp:337.)
  Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
INFO:ym:Train Epoch: 1 [0%]
INFO:ym:[2.4847304821014404, 3.50190806388855, 14.880080223083496, 21.834579467773438, 0.0, 2.532437801361084, 0, 9.99875e-05]
100it [02:04,  1.24s/it]
INFO:ym:====> Epoch: 1
0it [00:00, ?it/s]INFO:ym:Train Epoch: 2 [0%]
INFO:ym:[2.3833253383636475, 2.637192964553833, 3.6298272609710693, 17.414995193481445, 0.0, 1.991989254951477, 100, 9.99750015625e-05]
100it [01:59,  1.20s/it]
INFO:ym:====> Epoch: 2
![43e27b6d4f2d65334a0d29a3a55295d](https://github.com/RVC-Boss/GPT-SoVITS/assets/48117230/10a4ae72-a08e-4d18-b3a3-99763b15f739)

![3](https://github.com/RVC-Boss/GPT-SoVITS/assets/48117230/70709d8f-6ade-49a5-9215-49aedb45e06e)

**Comment by SapphireLab** [comment]
...你设置了总训练轮数为 2， 但是保存频率为 4 (意思就是你只训练了 2 轮还没有达到保存的轮数当然没有文件.
你可以调高总轮数, 或者调低保存频率.

**Comment by cjszhj** [comment]
谢谢。已解决。当时大意了，没仔细看。



> ...你设置了总训练轮数为 2， 但是保存频率为 4 (意思就是你只训练了 2 轮还没有达到保存的轮数当然没有文件. 你可以调高总轮数, 或者调低保存频率.

**Comment by Aelx007** [comment]
GPT_weights下没有文件，SoVITS_weights有文件，请大神指点一下

**Comment by Aelx007** [comment]
![948f7a73-6e95-4b11-92ed-91ae41f87b33](https://github.com/user-attachments/assets/94c02f19-7baa-4e77-bdab-a5e144853e8e)

**Comment by xiaodunmeng** [comment]
> GPT_weights下没有文件，SoVITS_weights有文件，请大神指点一下

兄弟解决了没啊，我是一样的问题

**Comment by lianqianmeng** [comment]
我个人不太熟悉这玩意，本地也出现类似问题，通过面向GPT对代码进行插桩，解决了问题，GPT给出的解释如下
@Aelx007 @xiaodunmeng 
太好了。那基本可以确认：**卡住点就是 `WN.forward()` 里的 fused 激活实现**，也就是原来的：

```python
acts = commons.fused_add_tanh_sigmoid_multiply(x_in, g_l, n_channels_tensor)
```

你现在能跑通，说明 `DataLoader`、`Encoder.pre`、`WN` 的卷积主体、`CFM` 至少不是最初卡死的根因。

建议保留这个稳定写法：

```python
in_act = x_in + g_l
acts = torch.tanh(in_act[:, : self.hidden_channels, :]) * torch.sigmoid(
    in_act[:, self.hidden_channels :, :]
)
```

后续可以做两件事：

1. **把调试 print 去掉**，只保留手写激活。
2. **继续保留长度保护**，尤其是 `mel_lengths = torch.clamp(mel_lengths, max=minn)`，避免后面 CFM/DiT 再因长度不一致出问题。

最终 `WN.forward()` 里核心部分建议这样：

```python
if g is not None:
    g = self.cond_layer(g)

for i in range(self.n_layers):
    x_in = self.in_layers[i](x)

    if g is not None:
        cond_offset = i * 2 * self.hidden_channels
        g_l = g[:, cond_offset : cond_offset + 2 * self.hidden_channels, :]
    else:
        g_l = torch.zeros_like(x_in)

    in_act = x_in + g_l
    acts = torch.tanh(in_act[:, : self.hidden_channels, :]) * torch.sigmoid(
        in_act[:, self.hidden_channels :, :]
    )

    acts = self.drop(acts)
    res_skip_acts = self.res_skip_layers[i](acts)

    if i < self.n_layers - 1:
        res_acts = res_skip_acts[:, : self.hidden_channels, :]
        x = (x + res_acts) * x_mask
        output = output + res_skip_acts[:, self.hidden_channels :, :]
    else:
        output = output + res_skip_acts
```

这次的问题结论可以写成一句：**`commons.fused_add_tanh_sigmoid_multiply` 在当前环境下导致 WN forward 卡住，替换为普通 PyTorch 的 `tanh * sigmoid` 后恢复正常。**

**Comment by lianqianmeng** [comment]
> > 我个人不太熟悉这玩意，本地也出现类似问题，通过面向GPT对代码进行插桩，解决了问题，GPT给出的解释如下 [@Aelx007](https://github.com/Aelx007) [@xiaodunmeng](https://github.com/xiaodunmeng) 太好了。那基本可以确认：**卡住点就是 `WN.forward()` 里的 fused 激活实现**，也就是原来的：
> > acts = commons.fused_add_tanh_sigmoid_multiply(x_in, g_l, n_channels_tensor)
> > 你现在能跑通，说明 `DataLoader`、`Encoder.pre`、`WN` 的卷积主体、`CFM` 至少不是最初卡死的根因。
> > 建议保留这个稳定写法：
> > in_act = x_in + g_l
> > acts = torch.tanh(in_act[:, : self.hidden_channels, :]) * torch.sigmoid(
> > in_act[:, self.hidden_channels :, :]
> > )
> > 后续可以做两件事：
> > 
> > 1. **把调试 print 去掉**，只保留手写激活。
> > 2. **继续保留长度保护**，尤其是 `mel_lengths = torch.clamp(mel_lengths, max=minn)`，避免后面 CFM/DiT 再因长度不一致出问题。
> > 
> > 最终 `WN.forward()` 里核心部分建议这样：
> > if g is not None:
> > g = self.cond_layer(g)
> > for i in range(self.n_layers):
> > x_in = self.in_layers[i](x)
> > ```
> > if g is not None:
> >     cond_offset = i * 2 * self.hidden_channels
> >     g_l = g[:, cond_offset : cond_offset + 2 * self.hidden_channels, :]
> > else:
> >     g_l = torch.zeros_like(x_in)
> > 
> > in_act = x_in + g_l
> > acts = torch.tanh(in_act[:, : self.hidden_channels, :]) * torch.sigmoid(
> >     in_act[:, self.hidden_channels :, :]
> > )
> > 
> > acts = self.drop(acts)
> > res_skip_acts = self.res_skip_layers[i](acts)
> > 
> > if i < self.n_layers - 1:
> >     res_acts = res_skip_acts[:, : self.hidden_channels, :]
> >     x = (x + res_acts) * x_mask
> >     output = output + res_skip_acts[:, self.hidden_channels :, :]
> > else:
> >     output = output + res_skip_acts
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 这次的问题结论可以写成一句：**`commons.fused_add_tanh_sigmoid_multiply` 在当前环境下导致 WN forward 卡住，替换为普通 PyTorch 的 `tanh * sigmoid` 后恢复正常。**


 或许这是因为我是5080卡，但没用对应的50系包导致的？
